src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -54,24 +79,79 @@ internal static unsafe string IPv4AddressToString(uint address)
         {
             const int MaxLength = 15;
             char* addressString = stackalloc char[MaxLength];
-            int offset = MaxLength;
 
-            FormatIPv4AddressNumber((int)((address >> 24) & 0xFF), addressString, ref offset);
-            addressString[--offset] = '.';
-            FormatIPv4AddressNumber((int)((address >> 16) & 0xFF), addressString, ref offset);
-            addressString[--offset] = '.';
-            FormatIPv4AddressNumber((int)((address >> 8) & 0xFF), addressString, ref offset);
-            addressString[--offset] = '.';
+            int charsWritten = IPv4AddressToStringHelper(address, addressString);
+
+            return new string(addressString, 0, charsWritten);
+        }
+
+        internal static unsafe bool IPv4AddressToString(uint address, Span<char> formatted, out int charsWritten)
+        {
+            const int MaxLength = 15;
+
+            if (formatted.Length < MaxLength)
+            {
+                charsWritten = 0;
+                return false;
+            }
+     
+            fixed (char* addressString = &formatted.DangerousGetPinnableReference())
+            {
+                charsWritten = IPv4AddressToStringHelper(address, addressString);
+            }
+
+            return true;
+        }
+
+        internal static unsafe int IPv4AddressToStringHelper(uint address, char* addressString)
```

Comment:
This no longer needs to use a pointer.  The pointer was here due to stackallocating, but now we can just write this in terms of `Span<char>`, making this method safer against possible overflow bugs.  The span-based caller can just pass in the target span.  And the string-based caller can still stackalloc the target space, but then rather than passing that in directly, it can create a span from it and pass in the span here.  That then also avoids the pinning that's been added in the span-based caller above.